From: George Beshers <gbeshers@redhat.com>
Date: Fri, 18 Dec 2009 22:23:36 -0500
Subject: [uv] React 2.6.32.y: isolcpus broken in 2.6.32.y kernel
Message-id: <4B2C00E8.5060900@redhat.com>
Patchwork-id: 22193
O-Subject: [RHEL6 PATCH] UV- React 2.6.32.y: isolcpus broken in 2.6.32.y kernel
Bugzilla: 548842
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

BZ548842:  UV- React 2.6.32.y: isolcpus broken in 2.6.32.y kernel

It turns out that others had seen the same problem and posted
the same patch.

Note that Ingo has signed off on it.  I will repost when it gets
into Linus' tree.

sched: Fix isolcpus boot option

Anton Blanchard wrote:

> We allocate and zero cpu_isolated_map after the isolcpus
> __setup option has run. This means cpu_isolated_map always
> ends up empty and if CPUMASK_OFFSTACK is enabled we write to a
> cpumask that hasn't been allocated.

I introduced this regression in 49557e620339cb13 (sched: Fix
boot crash by zalloc()ing most of the cpu masks).

Use the bootmem allocator if they set isolcpus=, otherwise
allocate and zero like normal.

Reported-by: Anton Blanchard <anton@samba.org>
Signed-off-by: Rusty Russell <rusty@rustcorp.com.au>
Cc: peterz@infradead.org
Cc: Linus Torvalds <torvalds@linux-foundation.org>
Cc: <stable@kernel.org>
LKML-Reference: <200912021409.17013.rusty@rustcorp.com.au>
Signed-off-by: Ingo Molnar <mingo@elte.hu>
Tested-by: Anton Blanchard <anton@samba.org>

Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/kernel/sched.c b/kernel/sched.c
index 5cbffde..2d68d90 100644
--- a/kernel/sched.c
+++ b/kernel/sched.c
@@ -8080,6 +8080,7 @@ static cpumask_var_t cpu_isolated_map;
 /* Setup the mask of cpus configured for isolated domains */
 static int __init isolated_cpu_setup(char *str)
 {
+	alloc_bootmem_cpumask_var(&cpu_isolated_map);
 	cpulist_parse(str, cpu_isolated_map);
 	return 1;
 }
@@ -9610,7 +9611,9 @@ void __init sched_init(void)
 	zalloc_cpumask_var(&nohz.cpu_mask, GFP_NOWAIT);
 	alloc_cpumask_var(&nohz.ilb_grp_nohz_mask, GFP_NOWAIT);
 #endif
-	zalloc_cpumask_var(&cpu_isolated_map, GFP_NOWAIT);
+	/* May be allocated at isolcpus cmdline parse time */
+	if (cpu_isolated_map == NULL)
+		zalloc_cpumask_var(&cpu_isolated_map, GFP_NOWAIT);
 #endif /* SMP */
 
 	perf_event_init();
